yaobow/fileformats/src/mv3.rs

```diff
@@ -1,263 +1,141 @@
-use byteorder::{LittleEndian, ReadBytesExt};
-use common::read_ext::ReadExt;
+use binrw::BinRead;
 use serde::Serialize;
-use std::io::Read;
+use std::io::{Read, Seek};
+
+use crate::{
+    rwbs::{Quaternion, TexCoord, Vec3f},
+    utils::{SizedString, StringWithCapacity},
+};
+
+#[derive(BinRead, Debug, Serialize, Clone)]
+#[brw(little)]
+pub struct Mv3Material {
+    pub diffuse: [f32; 4],
+    pub ambient: [f32; 4],
+    pub specular: [f32; 4],
+    pub emissive: [f32; 4],
+    pub power: f32,
+}
 
-#[derive(Debug, Serialize, Clone)]
+#[derive(BinRead, Debug, Serialize, Clone)]
+#[brw(little)]
 pub struct Mv3Texture {
-    pub unknown: Vec<u8>, // size: 68
-    pub names: Vec<Vec<u8>>,
+    pub material: Mv3Material,
+    #[br(count = 4)]
+    pub names: Vec<SizedString>,
 }
 
-#[derive(Debug, Serialize, Clone)]
+#[derive(BinRead, Debug, Serialize, Clone)]
+#[brw(little)]
 pub struct Mv3Vertex {
+    #[br(map = |v: i16| -v)]
     pub x: i16,
     pub y: i16,
+    #[br(map = |v: i16| -v)]
     pub z: i16,
     pub normal_phi: i8,
     pub normal_theta: u8,
 }
 
-#[derive(Debug, Serialize, Clone)]
+#[derive(BinRead, Debug, Serialize, Clone)]
+#[brw(little)]
+#[brw(import { count: usize })]
 pub struct Mv3Frame {
     pub timestamp: u32,
+    #[br(count = count)]
     pub vertices: Vec<Mv3Vertex>,
 }
 
-#[derive(Debug, Serialize, Clone)]
+#[derive(BinRead, Debug, Serialize, Clone)]
+#[brw(little)]
 pub struct Mv3Triangle {
     pub indices: [u16; 3],
     pub texcoord_indices: [u16; 3],
 }
 
-#[derive(Debug, Serialize, Clone)]
-pub struct Mv3UnknownDataInMesh {
-    pub u: u16,
-    pub v: u16,
-}
-
-#[derive(Debug, Serialize, Clone)]
+#[derive(BinRead, Debug, Serialize, Clone)]
+#[brw(little)]
 pub struct Mv3Mesh {
-    pub unknown: u32,
+    pub id: u32,
     pub triangle_count: u32,
+    #[br(count = triangle_count)]
     pub triangles: Vec<Mv3Triangle>,
-    pub unknown_data_count: u32,
-    pub unknown_data: Vec<Mv3UnknownDataInMesh>,
-}
-
-#[derive(Debug, Serialize, Clone)]
-pub struct Mv3TexCoord {
-    pub u: f32,
-    pub v: f32,
+    pub command_count: u32,
+    #[br(count = command_count)]
+    pub commands: Vec<u32>,
```

Comment:
learned from 0x7c13/Pal3.Unity

Reply:
is this acceptable?

Reply:
It will be great if I'm wrong but my impression is that Pal3.Unity references the leaked source code of the original PAL3 game... I'm still not feeling comfortable to do so, as referencing the leaked source code can make the project a derivative work of the original project (which is Commercial-licensed, of course). Referencing the source code of Pal3.Unity can be the same thing.
I'm still debating with myself on this topic (and discussions are welcomed). Can we remove the insights from Pal3.Unity for now and only keep the `binrw` rewrite part?

Reply:
Got it. You are really doing a hard but great job!  Respect to you! 😄
I'll drop those code as you suggested.

Reply:
Thanks!